### PR TITLE
VRF based leader election groundwork

### PR DESCRIPTION
**Pull request by HoOngEe**
In this pr, I implemented

* VRF sortition algorithm
* VRF hash generation by private keys in EngineSigner
* Introduction of PriorityMessage type in Tendermint

### CODE REVIEW

File: core/src/consensus/sortition/binom_cdf.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+use statrs::distribution::{Binomial, Univariate};
+
+/// Determine in which the lottery value is included walking along the
+/// binomial cumulative distribution function.
+///
+/// # Arguments
+///
+/// * `p`   - probability parameter in the binomial distribution
+/// * `n`   - trial prameter in the binomial distribution
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/binom_cdf.rs]: trial parameter

File: core/src/consensus/sortition/binom_cdf.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+/// * `lottery` - the target lottery value in the closed interval [0, 1]
+/// * `voting_power` - determines the number of walks in cdf, the maximum possible number of walks is n in the binomial distribution
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/binom_cdf.rs]: determines the number of walks in cdf, where the maximum possible number of walks is n in the binomial distribution

File: core/src/consensus/sortition/binom_cdf.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+///
+/// # Returns
+///
+/// * The j value where the lottery value is included in the interval [cdf(j - 1), cdf(j)]
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/binom_cdf.rs]: The j value that includes the lottery value in the interval

File: core/src/consensus/sortition/lot.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/lot.rs]: This file is a part of CodeChain.

File: core/src/consensus/sortition/lot.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use rug::{integer::Order, Integer, Rational};
+
+use super::binom_cdf::binomial_cdf_walk;
+
+/// Draws lots based on the vrf sortition algorithm. It divides the vrf_output by the
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/lot.rs]: what are lots? Are they just empty spaces? I don't know what they are, so I want to know.

File: core/src/consensus/sortition/lot.rs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+/// maximum value in the hash space and maps the result to a rational number.
+///
+/// # Arguments
+/// * `voting_power`    - voting power of the player who draws lots
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/lot.rs]: spacing error?

File: core/src/consensus/sortition/mod.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/mod.rs]: same correction here

File: core/src/consensus/sortition/vrf_sortition.rs [hunk 8]
```diff
@@ -0,0 +1,245 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/vrf_sortition.rs]: same

File: core/src/consensus/sortition/binom_cdf.rs [hunk 9]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/binom_cdf.rs]: in hopes that

File: core/src/consensus/sortition/lot.rs [hunk 10]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/lot.rs]: same

File: core/src/consensus/sortition/mod.rs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/mod.rs]: same here

File: core/src/consensus/sortition/vrf_sortition.rs [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
```
Comment (ScarletBlue) [inline core/src/consensus/sortition/vrf_sortition.rs]: same here

File: core/src/consensus/sortition/binom_cdf.rs [hunk 13]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Kodebox, Inc.
+// This file is part of CodeChain.
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/sortition/binom_cdf.rs]: This sentence is a fixed part of GPL license statement.

File: core/src/consensus/sortition/binom_cdf.rs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use statrs::distribution::{Binomial, Univariate};
+
+/// Determine in which the lottery value is included walking along the
```
[the remark this answers is not in the document]
Comment (HoOngEe) [reply core/src/consensus/sortition/binom_cdf.rs]: resolved

File: core/src/consensus/sortition/binom_cdf.rs (hunk 1 above)
Comment (HoOngEe) [reply core/src/consensus/sortition/binom_cdf.rs]: fixed.

File: core/src/consensus/sortition/binom_cdf.rs (hunk 2 above)
Comment (HoOngEe) [reply core/src/consensus/sortition/binom_cdf.rs]: fixed

File: core/src/consensus/sortition/binom_cdf.rs (hunk 3 above)
Comment (HoOngEe) [reply core/src/consensus/sortition/binom_cdf.rs]: fixed

File: core/src/consensus/sortition/lot.rs (hunk 5 above)
Comment (HoOngEe) [reply core/src/consensus/sortition/lot.rs]: I intended to express "제비뽑기".

**Comment by HoOngEe** [comment]
The build task fails in mac OS-10.14 but the openssl will not be used in the future so I merged it.

**Pull request merged**